### DISCUSSION TITLE
Bootstrap from existing repository that has flux-system manually?

### DISCUSSION THREAD

**Discussion by jasonumiker** [fluxcd/flux2#1777] [Q&A]
Hi,
I have been using Flux v2 in an environment where I tear down clusters and recreate them (mainly testing cluster provisioning/management automation as well as the ability to recover through recreation atm.)

My question is that the Flux manifests are already in the repo's flux-system folder - I can apply them via kubectl apply -k to put Flux (mostly) back. So, if I am recreating "the same" cluster what is the point of them being in the repo if `flux bootstrap` recreates them?

If I wanted to manually bootstrap using them what would be the process (recreate the secret to talk to GitHub etc.)?

**Comment by stefanprodan** [comment]
> If I wanted to manually bootstrap using them what would be the process (recreate the secret to talk to GitHub etc.)?

Please check out this example: https://fluxcd.io/docs/use-cases/azure/#flux-installation-for-azure-devops

**Reply by nikaro** [reply]
The `bootstrap` command create a new commit each time. Is there a way to tell use this repository as it is?

**Reply by stefanprodan** [reply]
It only creates a commit if the files in Git have diverged.

**Reply by nikaro** [reply]
Oh ok, that's the case because I run `ymlfmt` on the files of the repo 😅 Thanks !

**Reply by ppaz** [reply]
> It only creates a commit if the files in Git have diverged.

Hey sorry for chipping in, I want to wrap my head around Flux as newbie in subject. I hope it's also right place to start such discussion especially when I noticed similar subject.

I've noticed (which was super annoying) and did no understood why but my flux installation using terraform provider is doing commit almost always, like 9 out of 10 cases. Each time I destroy resources OR delete minikube and start fresh with new `terraform apply`

When I compared those `gotk-components.yaml`, it's because order of components has changed.

![image](https://github.com/user-attachments/assets/bf8f7dec-d150-4850-8e1e-8fc9cd47f82c)

Is that expected behavior? To be fair I am assuming, that I am doing something wrong, if not then I would expect to flux bootstrap to produce consistent end deterministic output everytime...

It's my config of `flux_bootstrap_git` and I am using `terraform` 1.5.7 and `flux` 2.3.0, running on Windows.
```
resource "flux_bootstrap_git" "fluxcd" {
  embedded_manifests = true
  path               = "clusters/local-minikube"
}
```

I've checked resource definition one by one, and the only difference was the sequence of components. I've even tried to list all components by name (but of course it's a unordered set). But I wanted to keep the same sequence as in `flux install`
```
resource "flux_bootstrap_git" "fluxcd" {
  embedded_manifests = true
  components         = toset([ "source-controller", "kustomize-controller", "helm-controller", "notification-controller" ])
  path               = "clusters/local-minikube"
}
```

What is weird is the fact that each run of terraform plan or terraform apply always gives me the same sequence of components:
```
      + components           = [
          + "helm-controller",
          + "kustomize-controller",
          + "notification-controller",
          + "source-controller",
        ]
```

Just FYI, `flux install --export` generates such sequence in manifest `Components: source-controller,kustomize-controller,helm-controller,notification-controller`.
![image](https://github.com/user-attachments/assets/3b7f4fc6-3de9-4332-9b9e-6e81d05fc170)

But apparently `flux terraform provider` has different or ?random? sequence, in most cases I observe manifest to have `Components: helm-controller,kustomize-controller,notification-controller,source-controller`

It was first spotted when I prepared manifests by `flux install` but then bootstrapped with terraform provider.

Any help on this?

**Reply by stefanprodan** [reply]
I guess Terraform stores the list internally as a Go map, and iterations over maps are random. Please open an issue in the terraform provider repo to track this.

**Reply by turbosheep44** [reply]
just came across this behavior as well using OpenTofu and the `fluxcd/flux` provider to bootstrap a cluster from an existing repo. @ppaz did you every end up creating an issue for this in the terraform provider repo, if so a link would be great as I can't find it!
